Add performance - scalability page (#3657)
istio/istio.io: content/docs/concepts/performance-and-scalability/index.md

---
title: Performance and Scalability

> description: Introduces Performance and Scalability for Istio components.

Reviewer: Don't capitalize Performance and Scalability. How about:

Introduces Istio's performance and scalability methodology.

keywords: [performance,scalability,scale,benchmarks]
---

Istio makes it easy to create a network of deployed services with rich routing,
 load balancing, service-to-service authentication, monitoring, and more. It
 does so without any changes to the application code. Istio strives to provide
 these benefits with minimal resource overhead. Istio also aims to support very
 large meshes with high request rates while adding minimal latency.

## Data plane performance

> Istio consists of components that run on the data plane and components that

Reviewer: I think that should be "in", not "on". Maybe enumerate the components in parenthesis for either planes?

 program the data plane, i.e. control plane. Data plane and control plane have
 distinct performance concerns.

### Latency

Istio injects a sidecar proxy on the data path therefore latency is an important
 consideration. Istio adds an auth and mixer filter to the proxy. Every
 additional filter adds to the path length inside the proxy and affects latency.

> Istio proxy collects raw telemetry data after the response is sent back to the

Reviewer: The Istio proxy
the response -> a response

 client. Time spent collecting raw telemetry for a request does not contribute

>  to the total time taken to complete that request. However, since the worker

Reviewer: Rephrase without using "thread", unecessaarily geeky.

 thread is busy doing this work, it will not start work on the next request
 immediately. This adds to the queue wait time of the next request affecting
 average and tail latencies. The actual tail latency depends on the traffic pattern.

Inside the mesh, a request traverses the client side proxy and then the server

>  side proxy. Two proxies on the data path add about `10ms` to the p99 latency at `1k rps`.

Reviewer: client-side
server-side

Reviewer 2: I'm a bit torn on whether or not we should be reporting the times for 2 proxies or 1 proxy. I understand the frankness of quoting numbers for 2 proxies...but I'm not sure people really think about that.  When they say "How much latency does the proxy add," I think they are thinking about 1 trip through the proxy.  WDYT?

Author: Question is how much latency does the mesh add usually. But PMs can decide :-)

Author: I added both numbers.

The proxy does not normally buffer data passing through it, therefore memory
 consumption is not affected by the request rate.

Data plane performance depends on many factors.

* Number of client connections
* Target request rate
* Request size and Response size
* Enabling `mTLS`
* Number of proxy worker threads
* Protocol
* Host cores

Latency, throughput and proxy CPU are measured as a function of the above parameters.

## Control plane performance

Pilot configures sidecar proxies based on user authored configuration and the current
 state of the system. In a Kubernetes environment services, deployments, and endpoints
 are the ambient configuration and state. Istio configuration like `gateway` and
 `virtual service` is the user authored portion of the configuration.
 Pilot processes the combined configuration and state to produce configuration for the proxies.

Control plane supports thousands of services, spread across thousands of pods with a
 similar number of user authored virtual services and other configuration.
 CPU and memory requirements for Pilot scale with the amount of configuration and state.

>  The CPU consumption also scales with the rate of change of deployments and the

Reviewer 3: Avoid listing factors in a paragraph, opt instead for lists:

```suggestion
The CPU consumption scales with the following factors:

- The rate of deployment changes.
- The rate of configuration changes.
- The number of proxies connecting to Pilot.

```

>  rate of change of configuration.

Reviewer 3: ```suggestion
```

> Lastly, the number of proxies connecting to a pilot affect the pilot CPU,

Reviewer 3: ```suggestion
```

>  however this part is inherently horizontally scalable.

Reviewer 3: You can add the horizontal scale behavior to the last bullet of the list.

When [namespace isolation](/docs/reference/config/networking/v1alpha3/sidecar/) is enabled,

>  a single pilot instance can support 1000 services, 2000 sidecars with 1 vCPU and 1.5 GB of memory.

Reviewer 3: ```suggestion
 a single Pilot instance can support 1000 services, 2000 sidecars with 1 vCPU and 1.5 GB of memory.
```

>  You may increase the number of pilots to reduce the amount of time it takes for configuration

Reviewer 3: ```suggestion
 You can increase the number of Pilot instances to reduce the amount of time it takes for the configuration
```

>  to reach all the proxies.

Reviewer 3: ```suggestion
 to reach all proxies.
```

## Release 1.1 performance summary

> **Mesh Traffic:** The total traffic in the mesh; includes ingress, egress, and all traffic going

Reviewer 3: Why are you including this definition here? This seems entirely out of place. I would include it as part of the intro.

 into and out of every pod.

### Data plane latency

> The default configuration of Istio 1.1 adds `10ms` to the p99 latency over the baseline.

Reviewer 3: ```suggestion
The default configuration of Istio 1.1 adds `10 ms` to the 99th percentile latency of the data plane over the baseline.
```

> The results were obtained by [Istio benchmarks](https://github.com/istio/tools/tree/master/perf/benchmark)

Reviewer 3: ```suggestion
We obtained these results using the [Istio benchmarks](https://github.com/istio/tools/tree/master/perf/benchmark)
```

>  for `http/1.1` protocol, `1kB` payload at `1k rps` using 16 client connections.

Reviewer 3: ```suggestion
 for the `http/1.1` protocol, with a `1 kB` payload at `1000 rps` using 16 client connections.
```

{{< image width="90%" ratio="75%"
    link="latency.svg?sanitize=true"
    alt="P99 latency vs client connections"
    caption="P99 latency vs client connections"
>}}

### Data plane resources

> Istio proxy uses 0.5 vCPU per 1000 rps and `istio-telemetry` uses 0.5 vCPU per 1000 mesh rps.

Reviewer 3: ```suggestion
The proxy uses 0.5 vCPU per 1000 rps and the `istio-telemetry` service uses 0.5 vCPU per 1000 mesh-wide rps.
```

### Control plane

> Istio pilot needs 1 vCPU and 1.5 GB of memory to support 1000 services and 2000 sidecars.

Reviewer 3: ```suggestion
Pilot uses 1 vCPU and 1.5 GB of memory to support 1000 services and 2000 sidecars.
```

keywords: [performance,scalability,scale,benchmarks]
---

Istio makes it easy to create a network of deployed services with rich routing,
load balancing, service-to-service authentication, monitoring, and more - all
without any changes to the application code. Istio strives to provide
these benefits with minimal resource overhead and aims to support very
large meshes with high request rates while adding minimal latency.

The Istio data plane components, the Envoy proxies and Mixer for headers only, handle
data flowing through the system. The Istio control plane components, Pilot, Galley and Citadel,
configure the data plane. The data plane and control plane have distinct
performance concerns.

## Data plane performance

Data plane performance depends on many factors, for example:

- Number of client connections
- Target request rate
- Request size and Response size
- Enabling `mTLS`
- Number of proxy worker threads
- Protocol

> - CPU cores

Reviewer 4: Is CPU cores a factor? Of course, the more cores you have, the better the performance, but this is a generic thing.  Shouldn't you mention mixer filter enabled/disabled and number of  policies as factors.

from the Kubernetes environment and the user-authored configuration.

The control plane supports thousands of services, spread across thousands of pods with a
similar number of user authored virtual services and other configuration objects.
Pilot's CPU and memory requirements scale with the amount of configurations and possible system states.
The CPU consumption scales with the following factors:

- The rate of deployment changes.
- The rate of configuration changes.
- The number of proxies connecting to Pilot.

however this part is inherently horizontally scalable.

When [namespace isolation](/docs/reference/config/networking/v1alpha3/sidecar/) is enabled,
a single Pilot instance can support 1000 services, 2000 sidecars with 1 vCPU and 1.5 GB of memory.
You can increase the number of Pilot instances to reduce the amount of time it takes for the configuration
to reach all proxies.

### Data plane latency

The default configuration of Istio 1.1 adds `10 ms` to the 99th percentile latency of the data plane over the baseline.
We obtained these results using the [Istio benchmarks](https://github.com/istio/tools/tree/master/perf/benchmark)

> for the `http/1.1` protocol, with a `1 kB` payload at `1000 rps` using 16 client connections.

Reviewer 4: Maybe specify the number of proxy workers and mTLS was disable for this.

{{< image width="90%" ratio="75%"
    link="latency.svg?sanitize=true"
    alt="P99 latency vs client connections"
    caption="P99 latency vs client connections"
>}}

- `baseline` Client pod directly calls the server pod, no sidecars are present.
- `server-sidecar` Only server sidecar is present.
- `both-sidecars` Client and server sidecars are present. This is a default case inside the mesh.
- `nomixer-both` Same as **both-sidecars** without the mixer filter.
- `nomixer-server` Same as **server-sidecar** without the mixer filter.

### Data plane resources

> The proxy uses 0.5 vCPU per 1000 rps and the `istio-telemetry` service uses 0.5 vCPU per 1000 mesh-wide rps.

Reviewer 4: I believe we were getting around 0.7 vCPU / 1000 rps for telemetry.

> ### Control plane

Reviewer 4: maybe "Control plane resources"

---

Istio makes it easy to create a network of deployed services with rich routing,
load balancing, service-to-service authentication, monitoring, and more - all
without any changes to the application code. Istio strives to provide
these benefits with minimal resource overhead and aims to support very
large meshes with high request rates while adding minimal latency.

The Istio data plane components, the Envoy proxies, handle data flowing through
the system. The Istio control plane components, Pilot, Galley and Citadel, configure
the data plane. The data plane and control plane have distinct performance concerns.

## Performance summary for Istio {{< istio_release_name >}}

The [Istio load tests](https://github.com/istio/tools/tree/master/perf/load) mesh consists
of **1000** services and **2000** sidecars with `70k` mesh-wide requests per second (`rps`).
After running the tests using Istio {{< istio_release_name >}}, we get the following results:

- The Envoy proxy uses **0.6 vCPU** and **50 MB memory** per `1k rps` going through the proxy .
- The `istio-telemetry` service uses **0.6 vCPU** per `1k` **mesh-wide** `rps`.
- Pilot uses **1 vCPU** and `1.5 GB` of memory.

> - The Envoy proxy adds `10 ms` to the 99th percentile latency.

Reviewer 5: @louiscryan argues that more folks look at 90th percentile latency in the Envoy community, and that we should be reporting at that level.

client. The time spent collecting raw telemetry for a request does not contribute
to the total time taken to complete that request. However, since the worker
is busy handling the request, the worker won't start handling the next request
immediately. This process adds to the queue wait time of the next request and affects
average and tail latencies. The actual tail latency depends on the traffic pattern.

Inside the mesh, a request traverses the client-side proxy and then the server-side
proxy. This two proxies on the data path add about `10ms` to the 99th percentile latency at `1k rps`.
The server-side proxy alone adds `6ms` to the 99th percentile latency.

### Latency for Istio {{< istio_release_name >}}

The default configuration of Istio 1.1 adds `10 ms` to the 99th percentile latency of the data plane over the baseline.
We obtained these results using the [Istio benchmarks](https://github.com/istio/tools/tree/master/perf/benchmark)
for the `http/1.1` protocol, with a `1 kB` payload at `1000 rps` using 16 client connections, 2 proxy workers and `mTLS` enabled.

{{< image width="90%" ratio="75%"
    link="latency.svg?sanitize=true"

>     alt="P99 latency vs client connections"

Reviewer 5: Same comment as before about 90%ile vs 99%ile - should we be reporting at 90%ile?

    caption="P99 latency vs client connections"
>}}

- `baseline` Client pod directly calls the server pod, no sidecars are present.
- `server-sidecar` Only server sidecar is present.

> - `both-sidecars` Client and server sidecars are present. This is a default case inside the mesh.

Reviewer 5: "This is a default case inside the mesh" seems to focus the attention on this line. If the state of the market is focusing on the latency added per proxy hop, then this will be compared against a single proxy hop worth of added latency.

I think we need to think about what we're most likely to be compared to. If it's single-proxy-hop stats at 90%ile, then we should present data that is appropriately compared to that.

however this part is inherently horizontally scalable.

When [namespace isolation](/docs/reference/config/networking/v1alpha3/sidecar/) is enabled,
a single Pilot instance can support 1000 services, 2000 sidecars with 1 vCPU and 1.5 GB of memory.
You can increase the number of Pilot instances to reduce the amount of time it takes for the configuration
to reach all proxies.

### Data plane latency

The default configuration of Istio 1.1 adds `10 ms` to the 99th percentile latency of the data plane over the baseline.
We obtained these results using the [Istio benchmarks](https://github.com/istio/tools/tree/master/perf/benchmark)
for the `http/1.1` protocol, with a `1 kB` payload at `1000 rps` using 16 client connections, 2 proxy workers and `mTLS` enabled.

{{< image width="90%" ratio="75%"
    link="latency.svg?sanitize=true"
    alt="P99 latency vs client connections"
    caption="P99 latency vs client connections"
>}}

- `baseline` Client pod directly calls the server pod, no sidecars are present.
- `server-sidecar` Only server sidecar is present.
- `both-sidecars` Client and server sidecars are present. This is a default case inside the mesh.

> - `nomixer-both` Same as **both-sidecars** without the mixer filter.

Reviewer 3: Please capitalize Mixer in the text to disambiguate the term for readers.
